Source/WebCore/platform/graphics/PathImpl.h

```diff
@@ -50,20 +50,18 @@ class PathImpl : public ThreadSafeRefCounted<PathImpl> {
 
     virtual Ref<PathImpl> copy() const = 0;
 
-    virtual void moveTo(const FloatPoint&) = 0;
-
-    virtual void addLineTo(const FloatPoint&) = 0;
-    virtual void addQuadCurveTo(const FloatPoint& controlPoint, const FloatPoint& endPoint) = 0;
-    virtual void addBezierCurveTo(const FloatPoint& controlPoint1, const FloatPoint& controlPoint2, const FloatPoint& endPoint) = 0;
-    virtual void addArcTo(const FloatPoint& point1, const FloatPoint& point2, float radius) = 0;
-
-    virtual void addArc(const FloatPoint&, float radius, float startAngle, float endAngle, RotationDirection) = 0;
-    virtual void addEllipse(const FloatPoint&, float radiusX, float radiusY, float rotation, float startAngle, float endAngle, RotationDirection) = 0;
-    virtual void addEllipseInRect(const FloatRect&) = 0;
-    virtual void addRect(const FloatRect&) = 0;
-    virtual void addRoundedRect(const FloatRoundedRect&, PathRoundedRect::Strategy) = 0;
-
-    virtual void closeSubpath() = 0;
+    void addSegment(PathSegment);
+    virtual void add(PathMoveTo) = 0;
+    virtual void add(PathLineTo) = 0;
+    virtual void add(PathQuadCurveTo) = 0;
+    virtual void add(PathBezierCurveTo) = 0;
+    virtual void add(PathArcTo) = 0;
+    virtual void add(PathArc) = 0;
+    virtual void add(PathEllipse) = 0;
+    virtual void add(PathEllipseInRect) = 0;
+    virtual void add(PathRect) = 0;
+    virtual void add(PathRoundedRect) = 0;
+    virtual void add(PathCloseSubpath) = 0;
```

Comment:
Is it okay to pass the PathSegment and the PathSegment data structures by value? PathRoundedRect has 12 floats and an enum.

Reply:
I would think so -- the PathSegment is only passed when it's a syntactic sugar over the individual calls. So no callframe should be generated, rather the switch should operate on the caller's value.
The individual PathSegment::Data types, especially small ones, should mostly be neutral when passed as a value:
For example, `add(PathMoveTo`) is two floats, where as `moveTo(const FloatPoint&)` passes the two simple values with a pointer indirection. Here the PathMoveTo variant should be a progression.

However, very big ones like `PathRoundedRect`, the ones that might have unread, redundant values, is a bit of a question mark. I would *imagine* that the effect is neutral, but I have not benchmarked it extensively.

Mostly I think the pass-by-value is progression if:
 - There is not many call layers that just forward the parameters (there is not)
 - There is not many variables that are passed redundantly and never read by the payload function (all except rounded rect variant)
 
At least if the problem is the latter, redundant values being passed, this would mean that the command is overloaded: two different methods in one. This is the case with "add rounded rect with platform strategy", "add rounded rect with webkit strategy". So a better, more consistent approach to fix that particular problem would be tahat these two could be split into two separate methods, if they mean different things.